specs/RuntimeSpec/Preparation.md

```diff
@@ -0,0 +1,235 @@
+limitations this process imposes on the contract authors.
+
+In order to provide high performance execution of the [function calls], the near validator utilizes
+ahead-of-time preparation of the contract WASM code. Today the contract code is prepared during the
+execution of the [`DeployContractAction`]. The results are then saved and later reported to the
+user when a [`FunctionCall`] is invoked.
+
+Note that only some parts of this document are normative. This document delves into implementation
+details, which may change in the future as long as the behavior documented by the normative
+portions of this book is maintained. The non-normative portions of this document will be called out
+as such.
+
+[function calls]: ./FunctionCall.md
+[`DeployContractAction`]: ./Actions.md#DeployContractAction
+[`FunctionCallAction`]: ./Actions.md#FunctionCallAction
+
+## High level overview
+
+**This section is not normative.**
+
+Upon initiation of a contract deployment, broadly the following operations will be executed:
+validation, instrumentation, conversion to machine code (compilation) and storage of the compiled
+artifacts in the account’s storage.
+
+Functions exported from the contract module may then be invoked through the mechanisms provided
+by the protocol. To common ways to call a function is by submitting a function call action onto
+chain or via a cross-contract call.
+
+Most of the errors that have occurred as part of validation, instrumentation, compilation, etc. are
+saved and reported when a `FunctionCallAction` is submitted. Deployment itself may only report
+errors relevant to itself, as described in the specification for [`DeployContractAction`].
+
+## Validation
+
+A number of limits are imposed on the WASM module that is being parsed:
+
+* The length of wasm code must not exceed `max_contract_size` genesis configuration parameter;
+* Module must be a valid module according to the WebAssembly core 1.0 specification (this means no
+  extensions such as multi value or simd; this limitation may be relaxed in the future).
+* The wasm module may contain no more than `1_000_000` distinct signatures;
+* The wasm module may contain no more than `1_000_000` function imports and local function
+  definitions;
+* The wasm module may contain no more than `100_000` imports;
+* The wasm module may contain no more than `100_000` exports;
+* The wasm module may contain no more than `1_000_000` global definitions and imports;
+* The wasm module may contain no more than `100_000` data segments;
+* UTF-8 strings comprising the wasm module definition (e.g. export name) may not exceed `100_000`
+  bytes each;
+* Function definitions may not specify more than `50_000` locals;
+* Signatures may not specify more than `1_000` parameters;
+* Signatures may not specify more than `1_000` results;
+* Tables may not specify more than `10_000_000` entries;
+* Module may not specify more than `1` table;
+* Module may not specify more than `1` memory;
```

Comment:
Hmm… are all the numbers above hardcoded or are they genesis parameters? If they're hardcoded, maybe we should at least add a comment alongside the place where they're hardcoded to remind people to edit this place too if the parameter were to change?

Reply:
Some of them come from the version of the wasm specification we use (e.g. multi-memory proposal is not enabled); some others are enforced by [the parser we use](https://github.com/bytecodealliance/wasm-tools/blob/main/crates/wasmparser/src/limits.rs).